src/scheduler/pbs_bitmap.c

```diff
@@ -0,0 +1,346 @@
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "pbs_bitmap.h"
+
+#define BYTES_TO_BITS(x) ((x) * 8)
+
+
+/**
+ * @brief allocate space for a pbs_bitmap (and possibly the bitmap itself)
+ * @param pbm - bitmap to allocate space for.  NULL to allocate a new bitmap
+ * @param num_bits - number of bits to allocate
+ * @return pbs_bitmap *
+ * @retval bitmap which was allocated
+ * @retval NULL on error
+ */
+pbs_bitmap *
+pbs_bitmap_alloc(pbs_bitmap *pbm, long num_bits)
+{
+	pbs_bitmap *bm;
+	unsigned long *tmp_bits;
+	long prev_longs;
+	
+	if(num_bits <= 0)
+		return NULL;
+	
+	if(pbm == NULL) {
+		bm = malloc(sizeof(pbs_bitmap));
+		if(bm == NULL)
+			return NULL;
+		bm->bits = NULL;
+		bm->num_longs = 0;
+	}
+	else 
+		bm = pbm;
+	
+	/* If we have enough unused bits available, we don't need to allocate */
+	if (bm->num_longs * BYTES_TO_BITS(sizeof(unsigned long)) >= num_bits) {
+		bm->num_bits = num_bits;
+		return bm;
+	}
+		
+		
+	prev_longs = bm->num_longs;
+	
+	bm->num_bits = num_bits;
+	bm->num_longs = num_bits / BYTES_TO_BITS(sizeof(unsigned long));
+	if (num_bits % BYTES_TO_BITS(sizeof(unsigned long)) > 0)
+		bm->num_longs++;
+	tmp_bits = calloc(bm->num_longs, sizeof(unsigned long));
+	if (tmp_bits == NULL) {
+		if(pbm == NULL) /* we allocated the memory */
+			pbs_bitmap_free(bm);
+		return NULL;
+	}
+	
+	if(bm->bits != NULL) {
+		int i;
+		for(i = 0; i < prev_longs; i++)
+			tmp_bits[i] = bm->bits[i];
+		
+		free(bm->bits);
+	}
+	bm->bits = tmp_bits;
+		
+	return bm;
+}
+
+/* pbs_bitmap destructor */
+void
+pbs_bitmap_free(pbs_bitmap *bm)
+{
+	if(bm == NULL)
+		return;
+	free(bm->bits);
+	free(bm);
+}
+
+/**
+ * @brief turn a bit on for a bitmap
+ * @param pbm - the bitmap
+ * @param bit - which bit to turn on
+ * @return nothing
+ */
+int
+pbs_bitmap_bit_on(pbs_bitmap *pbm, long bit)
+{
+	long n;
+	unsigned long b;
+	unsigned long f;
+	
+	if(pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits) {
+		if(pbs_bitmap_alloc(pbm, bit + 1) == NULL)
+			return 0;
+	}
+	
+	n = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	f = bit % BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << f;
+	
+	pbm->bits[n] |= b;
+	return 1;
+}
+
+/**
+ * @brief turn a bit off for a bitmap
+ * @param pbm - the bitmap
+ * @param bit - the bit to turn off
+ * @return nothing
+ */
+int
+pbs_bitmap_bit_off(pbs_bitmap *pbm, long bit)
+{
+	long n;
+	unsigned long b;
+	
+	if (pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits) {
+		if(pbs_bitmap_alloc(pbm, bit + 1) == NULL)
+			return 0;
+	}
+	
+	n = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << (bit % BYTES_TO_BITS(sizeof(unsigned long)));
```

Comment:
It might make the calling of this more readable, but the macro itself would not be.  I don't find it too bad.  I did hide some of it in BYTES_TO_BITS().  If you really want me to make this macro, I'll do it.  I'd rather not.

Reply:
That's ok, you can have it as is. It was just that you do a similar operation in multiple places.